### PR TITLE
Populate cost metadata for GATK tasks.

### PR DESCRIPTION
**Pull request by gbggrant**
Populates cost_metadata table for two gatk tasks

### CODE REVIEW

**Comment by gbggrant** [comment]
Passing integration test here: https://job-manager.dsde-prod.broadinstitute.org/jobs/2b9f4dc6-d058-4803-a583-5ab76fbd71e8

File: scripts/variantstore/wdl/GvsCreateFilterSet.wdl [hunk 1]
```diff
@@ -325,7 +333,12 @@ task ExtractFilterTask {
...
       -L ~{intervals} \
       --dataset-id ~{dataset_id} \
-      --project-id ~{query_project}
+      --project-id ~{query_project} \
+      --cost-observability-tablename ~{cost_observability_tablename} \
+      --call-set-identifier ~{call_set_identifier} \
+      --wdl-step GvsCreateFilterSet \
+      --wdl-call ExtractFilterTask \
+      --shard-identifier ~{intervals_name}
   >>>
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/GvsCreateFilterSet.wdl:342]: fully out of the scope of this pr, but we have so many params! I wonder if we could cut down on a lot of the fq ones--like would we ever not know what the alt-allele-table was called if we have the proj and dataset?

File: scripts/variantstore/wdl/GvsQuickstartIntegration.wdl [hunk 2]
```diff
@@ -61,6 +61,7 @@ workflow GvsQuickstartIntegration {
 
     call Unified.GvsUnified {
         input:
+            call_set_identifier = branch_name,
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/GvsQuickstartIntegration.wdl:64]: nice!

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java [hunk 3]
```diff
@@ -240,6 +266,24 @@ protected static VCFHeader generateVcfHeader(Set<String> sampleNames,
...
+    /**
+     * Enforce that if cost information is being recorded to the cost-observability-tablename then *all* recorded
+     * parameters are set
+     */
+    @Override
+    protected String[] customCommandLineValidation() {
+        final List<String> errors = new ArrayList<>();
+        if (projectID != null || datasetID != null || costObservabilityTableName != null || callSetIdentifier != null || wdlStep != null || wdlCall != null || shardIdentifier != null) {
+            if (projectID == null || datasetID == null || costObservabilityTableName == null || callSetIdentifier == null || wdlStep == null || wdlCall == null || shardIdentifier == null) {
+                errors.add("Parameters 'project-id', 'dataset-id', `cost-observability-tablename`, `call-set-identifier`, 'wdl-step', 'wdl-call', and 'shardIdentifier' must either ALL be set or ALL NOT BE set");
```
Comment (mcovarr) [inline src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java:278]: mix of single quotes and backticks around params

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeatures.java [hunk 4]
```diff
@@ -75,20 +69,67 @@ public class ExtractFeatures extends ExtractTool {
...
+    /**
+     * Enforce that if cost information is being recorded to the cost-observability-tablename then *all* recorded
+     * parameters are set
+     */
+    @Override
+    protected String[] customCommandLineValidation() {
+        final List<String> errors = new ArrayList<>();
+        if (projectID != null || datasetID != null || costObservabilityTableName != null || callSetIdentifier != null || wdlStep != null || wdlCall != null || shardIdentifier != null) {
+            if (projectID == null || datasetID == null || costObservabilityTableName == null || callSetIdentifier == null || wdlStep == null || wdlCall == null || shardIdentifier == null) {
+                errors.add("Parameters 'project-id', 'dataset-id', `cost-observability-tablename`, `call-set-identifier`, 'wdl-step', 'wdl-call', and 'shardIdentifier' must either ALL be set or ALL NOT BE set");
```
Comment (mcovarr) [inline src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeatures.java:118]: here too

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeaturesEngine.java [hunk 5]
```diff
@@ -133,16 +133,26 @@ public void traverse() {
...
+        final StorageAPIAvroReaderAndBigQueryStatistics storageAPIAvroReaderAndBigQueryStatistics = BigQueryUtils.executeQueryWithStorageAPI(
                 featureQueryString,
                 SchemaUtils.FEATURE_EXTRACT_FIELDS,
                 projectID,
                 datasetID,
                 userDefinedFunctions,
                 useBatchQueries,
                 cleanQueryLabels);
+        this.bigQueryQueryByteScanned = storageAPIAvroReaderAndBigQueryStatistics.queryStatistics.getTotalBytesProcessed();
+        this.storageAPIByteScanned = storageAPIAvroReaderAndBigQueryStatistics.storageAPIAvroReader.getEstimatedTotalBytesScanned();
```
Comment (mcovarr) [inline src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeaturesEngine.java:145]: nit: bytes (plural) scanned for both of these variable names

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/common/CostObservability.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                setInitialIntervalMillis(2000).
+                setMaxIntervalMillis(30000).
+                setMultiplier(2).
+                setRandomizationFactor(0.5).
+                build();
+        int retryCount = 0;
+        int maxRetries = 3;
+
+        while (true) {
+            // This uses the _default stream since it (a) commits immediately and (b) doesn't count
```
Comment (RoriCremer) [inline src/main/java/org/broadinstitute/hellbender/tools/gvs/common/CostObservability.java:68]: what does the single _ mean?

**Review by mcovarr — APPROVED** [review]

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java [hunk 7]
```diff
@@ -371,6 +415,21 @@ public void traverse() {
...
      }
 
+    @Override
+    public Object onTraversalSuccess() {
+        if (costObservabilityTableName != null) {
+            CostObservability costObservability = new CostObservability(projectID, datasetID, costObservabilityTableName);
+            // TODO - is this needed?
+//            costObservability.writeCostObservability(callSetIdentifier, wdlStep, wdlCall, shardIdentifier,
+//                    new Date(), new Date(), "BigQuery Query Scanned",
+//                    sampleList.getBigQueryQueryByteScanned() + engine.getBigQueryQueryByteScanned());
```
Comment (RoriCremer) [inline src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java:425]: I dont think we still have a query cost for the extractCohort step (just the prepare step before it)
We used to, but now the querying is done as part of the prepare step and for the extract, we just grab everything from the prepare tables with the readAPI

**Comment by rsasch** [comment]
Is there a way you can use/update the GvsQuickstartIntegration.[AssertCostIsTrackedAndExpected](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsQuickstartIntegration.wdl#L196) task to check for these additional tracked costs?

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/common/CostObservability.java (hunk 6 above)
Comment (gbggrant) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/common/CostObservability.java:68]: Really not sure - this was code that was largely refactored from 'LoadStatus'.  I will remove it.

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java (hunk 3 above)
Comment (gbggrant) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java:278]: Thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeaturesEngine.java (hunk 5 above)
Comment (gbggrant) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeaturesEngine.java:145]: 👍

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (gbggrant) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java:425]: Yeah - forgot to remove this after our discussion of this.

Comment (gbggrant) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohort.java:425]: Thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeatures.java (hunk 4 above)
Comment (rsasch) [reply src/main/java/org/broadinstitute/hellbender/tools/gvs/filtering/ExtractFeatures.java]: same — doesn't look like the params were treated in a consistent way

**Pull request merged**